### PR TITLE
Derive model-card provenance from release inputs

### PR DESCRIPTION
**Pull request by justinthelaw**
## What

- Require explicit merge lineage and ISO release-date inputs when preparing model artifacts.
- Derive the promoted checkpoint from validated lineage.
- Read the latest training loss/step and best validation loss from that checkpoint's `trainer_state.json`.
- Validate the base model, pipeline, checkpoint, numeric metrics, and date before replacing a payload.
- Remove the hard-coded checkpoint, losses, and release date from generated model cards.
- Add end-to-end and negative regression coverage.

## Why

The model-card generator currently publishes one historical checkpoint, loss pair, and release date for every future run. A new model can therefore be released with stale or false provenance even when its artifacts are different.

## Validation

- Targeted provenance tests: 5/5 passed.
- Available unittest suite: 8/8 passed.
- Markdown lint, Python AST/tabnanny, CLI help smoke, and `git diff --check` passed.

### CODE REVIEW

File: ml/profile-qa/profile_qa/prepare_hf_artifacts.py [hunk 1]
```diff
@@ -353,6 +526,13 @@ def _write_dataset_card(
 
 
 def prepare_model_payload(args: argparse.Namespace) -> Path:
+    lineage_file = getattr(args, "lineage_file", None)
+    if not isinstance(lineage_file, str) or not lineage_file.strip():
+        raise ValueError("model lineage is required; pass --lineage-file")
+    provenance = load_model_provenance(
+        Path(lineage_file),
+        getattr(args, "release_date", None),
+    )
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply ml/profile-qa/profile_qa/prepare_hf_artifacts.py]: Addressed in bb58d95a3b6f359019dc44e9f6ccb38e3910fab7: merge/export stages now carry deterministic lineage and content digests through the browser artifact, and preparation verifies the selected lineage, checkpoint, and artifact digest before replacing payloads.

File: ml/profile-qa/README.md [hunk 2]
```diff
@@ -69,7 +69,10 @@ PYTHONPATH=ml/profile-qa python -m profile_qa.train_lora --dataset ml/profile-qa
 PYTHONPATH=ml/profile-qa python -m profile_qa.evaluate --dataset ml/profile-qa/data/profile_qa.jsonl --model-id teapotai/teapotllm
 PYTHONPATH=ml/profile-qa python -m profile_qa.merge_adapter --adapter-model-id ml/profile-qa/checkpoints/teapot-profile-qa-lora/checkpoint-400 --output-dir ml/profile-qa/merged/teapot-profile-qa
 PYTHONPATH=ml/profile-qa ml/profile-qa/.venv-export/bin/python -m profile_qa.export_onnx --output-dir ml/profile-qa/onnx/candidate
-PYTHONPATH=ml/profile-qa python -m profile_qa.prepare_hf_artifacts --model-browser-dir ml/profile-qa/onnx/candidate/browser
+PYTHONPATH=ml/profile-qa python -m profile_qa.prepare_hf_artifacts \
+  --model-browser-dir ml/profile-qa/onnx/candidate/browser \
+  --lineage-file ml/profile-qa/merged/teapot-profile-qa/teapot_profile_qa_lineage.json \
+  --release-date YYYY-MM-DD
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply ml/profile-qa/README.md:89]: Addressed in bb58d95a3b6f359019dc44e9f6ccb38e3910fab7: the fine-tuning diagram and README now show evaluation reports, lineage/content digests, the browser artifact digest, and the explicit release-date input.

File: ml/profile-qa/profile_qa/prepare_hf_artifacts.py [hunk 3]
```diff
@@ -421,6 +602,16 @@ def main() -> int:
     parser.add_argument("--output-dir", default=str(PACKAGE_ROOT / "hf"))
     parser.add_argument("--model-repo-id", default=DEFAULT_MODEL_REPO_ID)
     parser.add_argument("--dataset-repo-id", default=DEFAULT_DATASET_REPO_ID)
+    parser.add_argument(
+        "--lineage-file",
+        required=True,
+        help="lineage JSON produced by profile_qa.merge_adapter",
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply ml/profile-qa/profile_qa/prepare_hf_artifacts.py:1382]: Addressed in bb58d95a3b6f359019dc44e9f6ccb38e3910fab7: fixed v5 report defaults are removed. Baseline and candidate reports must identify the correct revision/model digest, dataset, and split, and preparation validates them against the selected lineage.

**Comment by justinthelaw** [comment]
@codex review

File: ml/profile-qa/profile_qa/evaluate.py [hunk 4]
```diff
@@ -182,12 +261,11 @@ def main() -> int:
...
     if args.predictions_json:
         predictions = json.loads(Path(args.predictions_json).read_text(encoding="utf-8"))
-    elif args.model_id:
-        predictions = generate_predictions(args.model_id, records)
     else:
-        raise RuntimeError("provide --model-id or --predictions-json")
+        predictions = generate_predictions(args.model_id, records)
 
     report = score_predictions(records, predictions)
+    report["provenance"] = evaluation_provenance(args.model_id, args.split)
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply ml/profile-qa/profile_qa/evaluate.py]: Addressed in 2e6ba309aed7a702e805f04a3896680a9b54485a: reports now record the evaluated dataset SHA-256, and preparation hashes the selected dataset and rejects every baseline/validation/test report whose dataset digest differs. Stale-dataset preservation is covered.

File: ml/profile-qa/profile_qa/evaluate.py [hunk 5]
```diff
@@ -102,6 +110,73 @@ def _adapter_base_model_id(model_id: str) -> str | None:
...
+        )
+        adapter_digest = directory_sha256(model_path)
+        return {
+            "model_kind": "adapter",
+            "model_id": model_reference,
+            "model_sha256": adapter_digest,
+            "adapter_model_id": model_reference,
+            ADAPTER_DIGEST_FIELD: adapter_digest,
+            "base_model": PRIMARY_BASE_MODEL_ID,
+            "split": normalized_split,
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply ml/profile-qa/profile_qa/evaluate.py:295]: Addressed in 2e6ba309aed7a702e805f04a3896680a9b54485a: merge lineage and all report kinds now carry the pinned base-model revision; export and preparation require agreement with the current pin. Mismatched lineage/report revision tests were added.

File: ml/profile-qa/profile_qa/provenance.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    *,
+    source_lineage: dict[str, Any],
+    source_lineage_sha256: str,
+    stage: str,
+) -> Path:
+    """Write lineage metadata bound to an artifact directory's exact contents."""
+
+    lineage_path = artifact_dir / LINEAGE_FILENAME
+    payload = dict(source_lineage)
+    payload[SOURCE_LINEAGE_DIGEST_FIELD] = source_lineage_sha256
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply ml/profile-qa/profile_qa/provenance.py]: Addressed in 2e6ba309aed7a702e805f04a3896680a9b54485a: published lineage is constructed from an explicit public-field allowlist and uses a portable ASCII checkpoint label plus digest. Local canonical paths remain private, and unknown/injected or unsafe fields are rejected in tests.

**Comment by justinthelaw** [comment]
@codex review

File: ml/profile-qa/profile_qa/evaluate.py [hunk 7]
```diff
@@ -171,23 +268,32 @@ def generate_predictions(model_id: str, records: list[dict[str, Any]]) -> dict[s
...
-        predictions = generate_predictions(args.model_id, records)
     else:
-        raise RuntimeError("provide --model-id or --predictions-json")
+        predictions = generate_predictions(args.model_id, records)
 
     report = score_predictions(records, predictions)
+    report["provenance"] = evaluation_provenance(
+        args.model_id,
+        args.split,
+        dataset_sha256,
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply ml/profile-qa/profile_qa/evaluate.py]: Addressed in 724db5f7c7b046e04e1fabf2eda4d15f10818f0f: saved predictions use a provenance-bound envelope and are rejected unless model lineage/digest, canonical dataset digest, prompt digest, generation contract, split, and record IDs match exactly.

File: ml/profile-qa/profile_qa/evaluate.py [hunk 8]
```diff
@@ -102,6 +116,89 @@ def _adapter_base_model_id(model_id: str) -> str | None:
...
+        field=DATASET_DIGEST_FIELD,
+        source=Path("evaluation dataset"),
+    )
+    if model_id.rstrip("/") == PRIMARY_BASE_MODEL_ID:
+        return {
+            "model_kind": "baseline",
+            "model_id": PRIMARY_BASE_MODEL_ID,
+            BASE_MODEL_REVISION_FIELD: PRIMARY_BASE_MODEL_REVISION,
+            "base_model": PRIMARY_BASE_MODEL_ID,
+            DATASET_DIGEST_FIELD: normalized_dataset_sha256,
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply ml/profile-qa/profile_qa/evaluate.py:262]: Addressed in 724db5f7c7b046e04e1fabf2eda4d15f10818f0f: reports carry a digest of the exact ordered formatted instructions, including live public-profile context and prompt/history; preparation recomputes and validates it per split.

File: ml/profile-qa/profile_qa/export_onnx.py [hunk 9]
```diff
@@ -116,17 +204,38 @@ def get_browser_model_paths(quantized_dir: Path) -> list[Path]:
...
+        fp_dir,
+        source_lineage_sha256=lineage.sha256,
+        stage="onnx-fp",
+    )
+    for dtype, quantized_dir in quantized_dirs.items():
+        validate_artifact_lineage(
+            quantized_dir,
+            source_lineage_sha256=lineage.sha256,
+            stage=f"onnx-{dtype}",
+        )
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply ml/profile-qa/profile_qa/export_onnx.py]: Addressed in 724db5f7c7b046e04e1fabf2eda4d15f10818f0f: every quantized stage records and validates the exact full-precision parent digest; the browser marker retains the complete FP/int8/uint8 chain, stale same-lineage reuse is rejected, and fresh FP exports safely replace old trees.

File: ml/profile-qa/profile_qa/prepare_hf_artifacts.py [hunk 10]
```diff
@@ -4,22 +4,438 @@
...
+            f"evaluation report {report_path} evaluated model digest does not "
+            "match the selected merge lineage"
+        )
+
+
+def _load_evaluation_reports(
+    args: argparse.Namespace,
+    model_provenance: ModelProvenance,
+) -> EvaluationReports:
+    dataset_sha256 = file_sha256(Path(args.dataset))
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply ml/profile-qa/profile_qa/prepare_hf_artifacts.py]: Addressed in 724db5f7c7b046e04e1fabf2eda4d15f10818f0f: evaluation hashing and publication share one canonical LF JSONL serialization, so report provenance matches the bytes uploaded regardless of source whitespace or key ordering.

File: ml/profile-qa/profile_qa/prepare_hf_artifacts.py [hunk 11]
```diff
@@ -4,22 +4,477 @@
...
+    dataset_records = read_jsonl(Path(args.dataset))
+    dataset_sha256 = canonical_jsonl_sha256(dataset_records)
+    prompt_sha256_by_split = {
+        split: evaluation_prompt_sha256(_split_records(dataset_records, split))
+        for split in ("validation", "test")
+    }
+    baseline_path = Path(args.baseline_report)
+    validation_path = Path(args.validation_report)
+    test_path = Path(args.test_report)
+    baseline_test = _load_report(baseline_path)
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply ml/profile-qa/profile_qa/prepare_hf_artifacts.py:841]: Addressed in 724db5f7c7b046e04e1fabf2eda4d15f10818f0f: reports bind strict scoring and generation schemas plus normalized implementation/config digests, and preparation recomputes metrics from embedded predictions before rendering cards.

File: ml/profile-qa/profile_qa/prepare_hf_artifacts.py [hunk 12]
```diff
@@ -171,13 +637,14 @@ def _write_model_card(
 ## Training
 
 - Base model: `teapotai/teapotllm`
+- Base model revision: `{base_model_revision}`
 - Method: local LoRA/QLoRA continuation, no full fine-tune and no cloud training
-- Promoted checkpoint: `teapot-profile-qa-lora-v5/checkpoint-40`
+- Promoted checkpoint: `{promoted_checkpoint}`
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply ml/profile-qa/profile_qa/prepare_hf_artifacts.py:1072]: Addressed in 724db5f7c7b046e04e1fabf2eda4d15f10818f0f: model-card LoRA rank, alpha, dropout, targets, and related fields are derived from digest-validated `adapter_config.json`; unsupported per-module patterns/variants are rejected, and unverifiable quantization, batch, hardware, locality, and cloud claims are omitted.

**Comment by justinthelaw** [comment]
@codex review

File: ml/profile-qa/profile_qa/merge_adapter.py [hunk 13]
```diff
@@ -59,9 +79,19 @@ def main() -> int:
     (output_dir / LINEAGE_FILENAME).write_text(
         json.dumps(
             {
-                "adapter_model_id": args.adapter_model_id,
+                "schema_version": LINEAGE_SCHEMA_VERSION,
+                # This canonical path is private build metadata. Exported artifact
+                # markers retain only the portable checkpoint label and digest.
+                "adapter_model_id": str(adapter_path),
+                ADAPTER_CHECKPOINT_FIELD: adapter_checkpoint,
+                ADAPTER_DIGEST_FIELD: directory_sha256(adapter_path),
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply ml/profile-qa/profile_qa/merge_adapter.py:109]: Addressed in b457035b672d958a08d5f40686b4a44de4fd8912: merge now rejects equal, ancestor, descendant, and symlink-overlapping adapter/output paths before any write. Disjoint outputs are safely replaced so stale files cannot enter lineage; regressions cover all overlap cases, adapter preservation, and stale-output cleanup.

**Comment by justinthelaw** [comment]
@codex review

File: ml/profile-qa/profile_qa/provenance.py [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+    source_lineage: dict[str, Any],
+    source_lineage_sha256: str,
+    stage: str,
+    parent_artifact_sha256s: dict[str, str] | None = None,
+) -> Path:
+    """Write lineage metadata bound to an artifact directory's exact contents."""
+
+    lineage_path = artifact_dir / LINEAGE_FILENAME
+    payload = public_lineage_fields(source_lineage)
+    payload[SOURCE_LINEAGE_DIGEST_FIELD] = source_lineage_sha256
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply ml/profile-qa/profile_qa/provenance.py]: Fixed in `9131cba`. Artifact markers retain the schema-v1 `source_lineage_sha256` field, but now derive its value internally from the canonical sanitized public-lineage projection. Marker validation also requires the copied public fields to be complete and equal to the selected lineage, so a private `adapter_model_id` path neither appears in nor acts as a verifier for the published marker. Added private-path invariance, injection, and public-field tamper regressions.

File: ml/profile-qa/profile_qa/prepare_hf_artifacts.py [hunk 15]
```diff
@@ -4,28 +4,730 @@
...
+        model_provenance=model_provenance,
+        baseline=False,
+    )
+    _validate_report_scores(
+        promoted_validation,
+        report_path=validation_path,
+        records=_split_records(dataset_records, "validation"),
+    )
+    _validate_report_provenance(
+        promoted_test,
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply ml/profile-qa/profile_qa/prepare_hf_artifacts.py]: Fixed in `9131cba`. Model and dataset packaging now enforce complete report/provenance schemas, canonical baseline/adapter/merged IDs, report-slot roles, and exact equality of `(model_kind, model_id, model_sha256)` across promoted validation and test reports before any payload write. Added mixed adapter/merged, wrong-slot, missing/extra/private-field, and noncanonical-ID regressions.

File: ml/profile-qa/profile_qa/merge_adapter.py [hunk 16]
```diff
@@ -54,14 +95,26 @@ def main() -> int:
...
                 "base_model": PRIMARY_BASE_MODEL_ID,
-                "pipeline": "profile-qa-teapot-lora",
+                BASE_MODEL_REVISION_FIELD: PRIMARY_BASE_MODEL_REVISION,
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply ml/profile-qa/profile_qa/merge_adapter.py:111]: Fixed in `9131cba`. New PEFT adapters persist `PRIMARY_BASE_MODEL_REVISION` in `adapter_config.json` via `get_peft_model(..., revision=...)`. One centralized base-ID/revision validator now gates continuation, the resolved `--resume` checkpoint, adapter evaluation/generation, merge, and release packaging; ambiguous resume-plus-adapter selection is rejected. Added missing/stale revision and persistence/wiring regressions.

**Comment by justinthelaw** [comment]
@codex review
